### PR TITLE
DT-1059: add soft delete column for delete a case note new api endpoint

### PR DESCRIPTION
**Pull request by SimonMitchellMOJ**
added new column for soft delete and exclude soft deleted case notes from being returned

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java [hunk 1]
```diff
@@ -76,6 +95,14 @@
     @Column(columnDefinition = "serial", insertable = false, updatable = false)
     private Integer eventId;
 
+    @Builder.Default
+    @Column(nullable = false)
+    private Boolean deleted = Boolean.FALSE;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java:100]: could this be a primitive instead - why the nullable type?

File: src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNoteAmendment.java [hunk 2]
```diff
@@ -59,4 +74,8 @@
     @LastModifiedBy
     private String modifyUserId;
 
+    @Builder.Default
+    @Column(nullable = false)
+    private Boolean deleted = Boolean.FALSE;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNoteAmendment.java:79]: again why the nullable type here?

File: src/main/resources/application-local.yml [hunk 3]
```diff
@@ -4,3 +4,6 @@ spring:
       resourceserver:
         jwt:
           public-key-location: classpath:local-public-key.pub
+
+  flyway:
+    out-of-order: true
```
Comment (petergphillips) [inline src/main/resources/application-local.yml:9]: why was this required?

Comment (SimonMitchellMOJ) [reply src/main/resources/application-local.yml]: Removed!

File: src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java [hunk 4]
```diff
@@ -217,18 +217,69 @@ public void testGenerationOfEventId() {
...
+
+        final var caseNoteCountBefore = jdbcTemplate.queryForObject("SELECT COUNT(*) FROM offender_case_note_amendment", Integer.class);
+        assertThat(caseNoteCountBefore).isEqualTo(1);
+
+        repository.deleteOffenderCaseNoteAmendmentsByOffenderIdentifier("X2111XX");
+        final var deletedCaseNotes = repository.deleteOffenderCaseNoteByOffenderIdentifier("X2111XX");
+        assertThat(deletedCaseNotes).isEqualTo(1);
+
+        TestTransaction.flagForCommit();
+        TestTransaction.end();
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java:276]: don't you need to start a new transaction too here?

File: src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java [hunk 5]
```diff
@@ -217,18 +217,69 @@ public void testGenerationOfEventId() {
...
+        assertThat(repository.findById(persistedEntity.getId()).isEmpty()).isTrue();
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepositoryTest.java:277]: think can write
```
        assertThat(repository.findById(persistedEntity.getId())).isEmpty())
```
instead?

File: src/main/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepository.kt [hunk 6]
```diff
@@ -14,12 +14,24 @@ import java.util.*
 
 @Repository
 interface OffenderCaseNoteRepository : PagingAndSortingRepository<OffenderCaseNote, UUID>, JpaSpecificationExecutor<OffenderCaseNote> {
+
   fun findBySensitiveCaseNoteType_ParentType_TypeInAndModifyDateTimeAfterOrderByModifyDateTime(types: Set<String>?, createdDate: LocalDateTime?, page: Pageable?): List<OffenderCaseNote>
+
   fun findByModifyDateTimeBetweenOrderByModifyDateTime(fromDateTime: LocalDateTime, toDateTime: LocalDateTime): List<OffenderCaseNote>
 
+  @Query(value = "SELECT ocn.* FROM OFFENDER_CASE_NOTE ocn WHERE ocn.offender_case_note_id = ?1", nativeQuery = true)
+  fun findByIdIgnoreSoftDelete(offenderCaseNoteId: UUID): Optional<OffenderCaseNote>
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepository.kt:23]: bit naff this method since it is only used by tests.  Would be better to inject the `JdbcTemplate` or similar into the test and then can just run the sql there

File: src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java [hunk 7]
```diff
@@ -76,6 +95,13 @@
     @Column(columnDefinition = "serial", insertable = false, updatable = false)
     private Integer eventId;
 
+    @Builder.Default
+    private boolean softDeleted = false;
+
+    public void setSoftDeleted(final boolean softDeleted) {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java:101]: not sure you need this method if you use a `@SQLDelete` annotation?

File: src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java [hunk 8]
```diff
@@ -76,6 +95,13 @@
     @Column(columnDefinition = "serial", insertable = false, updatable = false)
     private Integer eventId;
 
+    @Builder.Default
+    private boolean softDeleted = false;
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/model/OffenderCaseNote.java:99]: maybe don't need this field?

File: src/main/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepository.kt [hunk 9]
```diff
@@ -14,12 +14,24 @@ import java.util.*
...
+
   @Modifying
   @Query("UPDATE OffenderCaseNote ocn SET ocn.offenderIdentifier = :newOffenderIdentifier WHERE ocn.offenderIdentifier = :oldOffenderIdentifier")
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/repository/OffenderCaseNoteRepository.kt:26]: wondering if we need to change this to native sql so works for soft deleted records too?
